packages/Button/src/Button.tsx

```diff
@@ -267,61 +292,126 @@ const baseStyle = css`
   }
 `;
 
-export default class Button extends Component {
-  static displayName = 'Button';
+interface SharedButtonProps {
+  variant: Variant;
+  size: Size;
+  className?: string;
+  children?: React.ReactNode;
+  disabled?: boolean;
+}
 
-  static propTypes = {
-    variant: PropTypes.oneOf(['default', 'primary', 'info', 'danger', 'dark']),
-    size: PropTypes.oneOf(['xsmall', 'small', 'normal', 'large']),
-    className: PropTypes.string,
-    children: PropTypes.node,
-    disabled: PropTypes.bool,
-    as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
-    href: PropTypes.string,
-  };
+interface LinkButtonProps extends HTMLElementProps<'a'>, SharedButtonProps {
+  href: string;
+}
+
+interface ButtonButtonProps
+  extends HTMLElementProps<'button'>,
+    SharedButtonProps {
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedButtonProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function usesCustomElement(
+  props: ButtonProps,
+): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function hasLinkRoot(
```

Comment:
```suggestion
function usesLinkElement(
```

Reply:
ah shoot, yeah good call

Reply:
side-note, I wish I could've just committed this from inside GitHub's UI, but alas... GitHub doesn't have TypeScript's understanding of renames...